### PR TITLE
Fix ExpandedValue sanitization on struct collection

### PR DESCRIPTION
**Pull request by theomagellan**
#### Description
`useExpandedValue` would resolve ExpandedValues after checking if a collection's type was "stringy." For collections of structs, `isStringy` always returns false, causing all ExpandedValues to be sanitized to their parsed values and breaking decoding of string fields if the parsed value was not `string`.

Now, `useExpandedValue` checks if it's dealing with a collection of structs; in which case, skip sanitization and rely on mapstructure's per-field decoding.

#### Testing

Added tests relying on `configopaque.MapList` which is an alias on the struct collection `[]Pair`

### CODE REVIEW

**Comment by theomagellan** [comment]
As discussed with @mx-psi, I'm kindly pinging @jade-guiton-dd to this PR!

**Comment by jade-guiton-dd** [comment]
Hello 👋 Sorry, even after reading the existing code and your PR, I still don't really understand what the problem was. Could you expand on:

> causing all ExpandedValues to be sanitized to their parsed values and breaking decoding of string fields if the parsed value was not string.

Reading [the code](https://github.com/open-telemetry/opentelemetry-collector/blob/fc5681aa926c2666a1e2b1e2866cc3ac10d93372/confmap/internal/conf.go#L226), it seems to me that an array of structs should already be kept as-is by the sanitization operation. Did I miss something?

**Comment by theomagellan** [comment]
> Reading [the code](https://github.com/open-telemetry/opentelemetry-collector/blob/fc5681aa926c2666a1e2b1e2866cc3ac10d93372/confmap/internal/conf.go#L226), it seems to me that an array of structs should already be kept as-is by the sanitization operation. Did I miss something?

The issue used to happen in cases like what this [test does](https://github.com/open-telemetry/opentelemetry-collector/pull/14413/changes#diff-c14c3e295433e7119f8301d6c7f327cec44da11567f16c6a54636dbd2d28a5ddR48).

When decoding an `ExpandedValue` inside a struct field in a collection, the code would decode the `ExpandedValue` as its `Value` field (of `any` type) because we were deeming that a struct type was not stringy.
This would break cases like mentioned above where we would, later on in the unmarshalling process, try to set the result of our decoded `ExpandedValue` inside a struct field that would have been decoded as stringy if the decoding operation would've been delayed.

Commenting the fix and running the added tests fails with the error:
```
FAIL: TestMapListWithExpandedValueIntValue (0.00s)
[...]
'headers[0].value' expected type 'configopaque.String', got unconvertible type 'int'
```
Since the `ExpandedValue` was decoded as its `Value` because `useExpandedValue` called `isStringyStructure` on the type of the collection (collection is `[]Pair`, so the type is `Pair`) which is not a stringy type.

Sorry if I wasn't clear enough before, I hope my explanations help!

**Comment by jade-guiton-dd** [comment]
Ah, it seems there was a misunderstanding on my part (`isStringyStructure` checks the type of the destination, whereas `sanitizeExpanded` switches over the type of the source). The PR makes sense to me now.

One thing I thought about is that, in the same way your PR delegates to `mapstructure` the task of inspecting the struct so we can more accurately tell when we're assigning `ExpandedValue`s to strings, have you tried completely deleting the `switch to.Kind() {` blocks, so that arrays/slices/maps all get mapped by `mapstructure` instead of the `sanitize` functions, and we just rely on the initial test to identify values to keep as strings? It's possible there's an edge case I'm not thinking of where that wouldn't work, but I feel like that would be worth a try to simplify the logic instead of making it more complex.

**Comment by theomagellan** [comment]
I tested it and it also solves the issue and doesn't break any other tests.
My understanding of the code was that these checks were here for performance improvements, as for cases like `[]string` we can avoid subsequent hook calls since we already know that the collection is only composed of stringy elements.

I do agree that removing the switch altogether would make the code simpler and would also resolve any edge cases like the one I brought up. What do you think?

**Comment by jade-guiton-dd** [comment]
Regardless of Pablo's intent 2 years ago, I suspect it's not actually an optimization: I think `mapstructure` will walk the sanitized slice/map returned by the hook and call it again on each contained items regardless. So I don't think removing the switch should cause a performance regression; I'd say it's an unconditional improvement.

However, I'm not 100% confident that this won't break someone somewhere (`mapstructure`'s behavior and the interactions between our hooks can be a bit arcane in my experience), so I think it may be prudent to introduce a beta feature gate and run the latter part of the function if the gate has been disabled. That way we have a quick solution if someone encounters breakage; and if we don't receive any complaints, we can stabilize the gate and fully remove the code in a few releases.

File: confmap/internal/featuregates.go [hunk 1]
```diff
@@ -12,3 +12,11 @@ var EnableMergeAppendOption = featuregate.GlobalRegistry().MustRegister(
 	featuregate.WithRegisterDescription("Combines lists when resolving configs from different sources. This feature gate will not be stabilized 'as is'; the current behavior will remain the default."),
 	featuregate.WithRegisterReferenceURL("https://github.com/open-telemetry/opentelemetry-collector/issues/8754"),
 )
+
+var DeferExpandedValueSanitizationOnStructCollection = featuregate.GlobalRegistry().MustRegister(
+	"confmap.deferExpandedValueSanitizationOnStructCollection",
+	featuregate.StageBeta,
+	featuregate.WithRegisterFromVersion("v0.144.0"),
+	featuregate.WithRegisterDescription("Disables early sanitization of ExpandedValue during config unmarshalling, allowing mapstructure to handle type conversion at the field level. Fixes decoding errors when environment variable values are parsed as non-string types (e.g., numbers, booleans) but need to be assigned to string fields."),
+	featuregate.WithRegisterReferenceURL("https://github.com/open-telemetry/opentelemetry-collector/pull/14413#issuecomment-3754949484"),
```
Comment (theomagellan) [inline confmap/internal/featuregates.go]: All other featuregates reference issues directly instead of PRs.
Should I create an issue or is this good enough? Sorry.

File: confmap/internal/decoder.go [hunk 2]
```diff
@@ -93,14 +93,26 @@ func useExpandValue() mapstructure.DecodeHookFuncType {
...
+		if !DeferExpandedValueSanitizationOnStructCollection.IsEnabled() {
+			switch to.Kind() {
+			case reflect.Array, reflect.Slice:
+				elemType := to.Elem()
+				// If the element type is a struct, don't sanitize at the collection level.
+				// Let mapstructure decode field-by-field so each ExpandedValue can be
+				// converted based on its target field type.
+				if elemType.Kind() == reflect.Struct {
+					return data, nil
+				}
```
Comment (theomagellan) [inline confmap/internal/decoder.go]: I still kept the original fix inside the old behavior since it's still required to fix the problem I encountered.
I'm happy to remove it if you think it's unnecessary.

File: confmap/internal/featuregates.go (hunk 1 above)
Comment (jade-guiton-dd) [reply confmap/internal/featuregates.go]: I think referencing a PR should be fine.

**Comment by mx-psi** [comment]
> Regardless of Pablo's intent 2 years ago, I suspect it's not actually an optimization: I think `mapstructure` will walk the sanitized slice/map returned by the hook and call it again on each contained items regardless. So I don't think removing the switch should cause a performance regression; I'd say it's an unconditional improvement.

To be clear although I don't remember exactly what I was thinking at that point I don't think my intent here was performance-related, it was probably more just an oversight/something behavior-related.

> However, I'm not 100% confident that this won't break someone somewhere (`mapstructure`'s behavior and the interactions between our hooks can be a bit arcane in my experience), so I think it may be prudent to introduce a beta feature gate and run the latter part of the function if the gate has been disabled. That way we have a quick solution if someone encounters breakage; and if we don't receive any complaints, we can stabilize the gate and fully remove the code in a few releases.

I agree with this, mapstructure is always arcane!

File: confmap/internal/decoder.go (hunk 2 above)
Comment (jade-guiton-dd) [reply confmap/internal/decoder.go]: Hm, I doubt the original fix would cause issues, but out of caution, I'd rather have disabling the feature gate return to the previous behavior, bug included. I figure that if someone's setup is broken by the new fix, I think there's a reasonable chance they will be broken by the original fix as well.

File: confmap/internal/featuregates.go (hunk 1 above)
Comment (jade-guiton-dd) [inline confmap/internal/featuregates.go]: Nitpick: I think some of the details could be abbreviated a bit to avoid overwhelming the output of the `featuregate` command.

```suggestion
var NewExpandedValueSanitizer = featuregate.GlobalRegistry().MustRegister(
	"confmap.newExpandedValueSanitizer",
	featuregate.StageBeta,
	featuregate.WithRegisterFromVersion("v0.144.0"),
	featuregate.WithRegisterDescription("Fixes some types of decoding errors where environment variables are parsed as non-string types but assigned to string fields."),
	featuregate.WithRegisterReferenceURL("https://github.com/open-telemetry/opentelemetry-collector/pull/14413"),
```

File: confmap/internal/e2e/maplist_expanded_test.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+	conf := confmap.NewFromStringMap(data)
+	var tc testConfig
+	err := conf.Unmarshal(&tc)
+	// This should work because useExpandValue detects the target is a string
+	require.NoError(t, err)
+	require.Equal(t, configopaque.String("8080"), tc.Value)
+}
+
+// TestStringyStructureWithExpandedValue tests the isStringyStructure path in useExpandValue
+func TestStringyStructureWithExpandedValue(t *testing.T) {
```
Comment (theomagellan) [inline confmap/internal/e2e/maplist_expanded_test.go:110]: With this new test, we should now cover all cases in both the new and old behavior.

File: confmap/internal/e2e/Makefile [hunk 4]
```diff
@@ -1 +1,4 @@
 include ../../../Makefile.Common
+
+# Override COVER_PKGS to include the parent package that this e2e module tests
+COVER_PKGS := go.opentelemetry.io/collector/confmap/internal,$(COVER_PKGS)
```
Comment (theomagellan) [inline confmap/internal/e2e/Makefile]: This seems to be needed since I am covering code in `confmap/internal` from the `e2e` module.
This feels out of scope of the PR to me and I would appreciate any suggestions on how we could handle this coverage issue!

Comment (mx-psi) [reply confmap/internal/e2e/Makefile]: I think we can do this in a separate PR if you are up to contributing it. In any case it's not a requirement for merging this PR (codecov is not a required check)

Comment (theomagellan) [reply confmap/internal/e2e/Makefile]: Thank you. I will revert this commit and open a new PR.
In the meantime, do you have any ideas why the changelog validate is unhappy? I don't really understand the issue here.

Comment (jade-guiton-dd) [reply confmap/internal/e2e/Makefile]: I think that's an issue from other changelogs on main, so no worries.

File: .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+change_type: 'bug_fix'
+
+# The name of the component, or a single word describing the area of concern, (e.g. receiver/otlp)
+component: pkg/confmap
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: 'fix ExpandedValue sanitization when parsed as non-string but assigned to string fields'
+
+# One or more tracking issues or pull requests related to the change
+issues: []
```
Comment (mx-psi) [inline .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]:
```suggestion
issues: [14413]
```

File: .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: 'bug_fix'
+
+# The name of the component, or a single word describing the area of concern, (e.g. receiver/otlp)
+component: pkg/confmap
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: 'fix ExpandedValue sanitization when parsed as non-string but assigned to string fields'
```
Comment (mx-psi) [inline .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]: Can we rewrite this as something end-user focused? End-users don't know what `ExpandedValue` is, so this release note is not going to be informational for them. 


Maybe...

```suggestion
note: 'Use string representation when mapping to string fields in arbitrary data structures'
```

(see other comment about subtext)

File: .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+# One or more tracking issues or pull requests related to the change
+issues: []
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:
```
Comment (mx-psi) [inline .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]:
```suggestion
subtext: |
   For example, a header can be set via an environment variable to a string that is parseable as a number, e.g. `1234`
```

File: .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml (hunk 5 above)
Comment (theomagellan) [reply .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]: I didn't know that issues could directly redirect to PRs, neat! Thank you.

File: .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml (hunk 6 above)
Comment (jade-guiton-dd) [reply .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]: I'm not sure an end user reading this will understand that this is a fix for a config decoding error that occurs in some edge cases. Maybe something like:

> Fix config decoding error when setting a string field to an interpolated value in some circumstances.

(And the subtext you suggested should help fill in the blanks on what "circumstances" those are)

Comment (theomagellan) [reply .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]: If I can chime in, I would lean towards something like
> Fix an issue where configs could fail to decode when using interpolated values in string fields.

This seems closer to where an end-user could encounter the issue and also clears out the "some circumstances" situation.
What do you think?

Comment (jade-guiton-dd) [reply .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]: Yes, that sounds much better. "Could" should be enough to indicate it's not systematic.

Comment (mx-psi) [reply .chloggen/fix-expandedvalue-sanitization-on-struct-collection.yaml]: Works for me, thanks

**Comment by mx-psi** [comment]
Can you solve the merge conflict? Once you have dealt with that I can merge it
